test: increase test-crypto.js strictness

Trott opened the pull request:
Confirm that `getCiphers()` contains no duplicates.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
test crypto tls

lpinca commented on test/parallel/test-crypto.js:
```diff
@@ -54,19 +54,18 @@ assert.throws(function() {
 }
 
 // Assume that we have at least AES-128-CBC.
-assert.notStrictEqual(0, crypto.getCiphers().length);
```
Any reason for removing this?

sam-github replied:
Without this, `getCiphers()` could return `[]`, and the test would pass, because the empty set is sorted and unique.

Trott replied:
> Without this, getCiphers() could return [], and the test would pass, because the empty set is sorted and unique.

The test will fail because the very next line requires that the array includes `aes-128-cbc` as an element. So it can't be length 0. So the zero-length check seems redundant to me.

Although looking more closely, I suppose it's possible that `getCiphers()` would return different arrays each time called. (That would be a bug, but that's the point of tests.) So I should probably assign the result to a variable before all the asserts. Maybe even throw in another call to `getCiphers()` at the end and confirm that the array it returns is identical to the one it returned in the first place.

sam-github replied:
That is reasonably compelling, though an extra assert to be explicit about the condition isn't bad. As you wish, I'm OK either way.

Trott replied:
Eh, well, two people asked "why did you take this out?" and it isn't hurting anything, so maybe best to leave it in after all. I'll put it and the others back.

sam-github commented on lib/tls.js:
```diff
@@ -22,9 +22,9 @@ exports.DEFAULT_CIPHERS =
-exports.getCiphers = internalUtil.cachedResult(() => {
-  return internalUtil.filterDuplicateStrings(binding.getSSLCiphers(), true);
-});
+exports.getCiphers = internalUtil.cachedResult(
```
Unrelated change. I don't object to cleaning up `lib/`'s use of rocket functions, but I think they should be standalone commits, ideally a standalone commit per file that makes every use in the file consistent (this might or might not be the only such usage, I can't tell from the diff, and the commit message doesn't even mention that there is any cleanup occuring) or just to the entire lib directory to clean up a specific usage aspect.

I also don't think commits called `test: ...` should touch lib, unless the lib change really clearly is something needed to support the test code (an example might be exporting an internal so that it can be directly used in a unit test)

Trott replied:
Yeah, I'll claim a combination of being tired when I did this, multi-tasking too much, and/or just being lazy. The change originally started in `tls.js` and morphed into fixing up the test. When the morphing was complete, I shouldn't have included this now-unrelated change. I'll remove it.

sam-github commented on test/parallel/test-crypto.js:
```diff
@@ -54,19 +54,18 @@ assert.throws(function() {
-  assert.deepStrictEqual(list, sorted);
+  // list is sorted and contains no duplicates
+  assert(list.every(
+    (val, index, array) => index === array.length - 1 || val < array[index + 1]
```
While it works as commented, this is a pretty obfuscated way of doing both a sorted test and a uniqueness test in a single statement. I prefer the approach in https://github.com/nodejs/node/blob/20efbffaf8c5f4af94da31c95ba90e77a2e00e08/test/parallel/test-process-getgroups.js#L30-L32, though in hind-sight, maybe that `unique()` function should not *also* be sorting. Maybe `unique()` should be in `test/common.js`?

Trott replied:
I'll rework it to try to make it a bit more clear what's going on.

Trott commented:
@sam-github I think I've addressed your comments. PTAL.

I'll want to expand this test a bit more as I think I've discovered an unexpected behavior, but one thing at a time...

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/5872/

italoacasas commented:
Landed https://github.com/nodejs/node/commit/e21126d20f10b319593b755312dea13cd39f048a
